iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/aggregation/UDAFAccumulator.java

```diff
@@ -0,0 +1,219 @@
+    this.configurations = new UDAFConfigurations(ZoneId.systemDefault());
+
+    List<String> childExpressionStrings =
+        childrenExpressions.stream()
+            .map(Expression::getExpressionString)
+            .collect(Collectors.toList());
+    beforeStart(
+        childExpressionStrings,
+        Collections.singletonList(childrenExpressionDataTypes),
+        attributes,
+        isInputRaw);
+  }
+
+  private void beforeStart(
+      List<String> childExpressions,
+      List<TSDataType> childExpressionDataTypes,
+      Map<String, String> attributes,
+      boolean isInputRaw) {
+    reflectAndValidateUDF(childExpressions, childExpressionDataTypes, attributes, isInputRaw);
+    configurations.check();
+  }
+
+  private void reflectAndValidateUDF(
+      List<String> childExpressions,
+      List<TSDataType> childExpressionDataTypes,
+      Map<String, String> attributes,
+      boolean isInputRaw) {
+    udaf = (UDAF) UDFManagementService.getInstance().reflect(functionName);
+    state = udaf.createState();
+
+    final UDFParameters parameters =
+        new UDFParameters(
+            childExpressions,
+            UDFDataTypeTransformer.transformToUDFDataTypeList(childExpressionDataTypes),
+            attributes);
+
+    // Only validate for raw input
+    // There is no need to validate for partial input
+    if (isInputRaw) {
+      try {
+        // Double validates in workers
+        udaf.validate(new UDFParameterValidator(parameters));
+      } catch (Exception e) {
+        onError("validate(UDFParameterValidator)", e);
+      }
+    }
+
+    try {
+      udaf.beforeStart(parameters, configurations);
+    } catch (Exception e) {
+      onError("beforeStart(UDFParameters, UDAFConfigurations)", e);
+    }
+  }
+
+  @Override
+  public void addInput(Column[] columns, BitMap bitMap) {
+    udaf.addInput(state, columns, bitMap);
+  }
+
+  @Override
+  public void addIntermediate(Column[] partialResult) {
+    checkArgument(partialResult.length == 1, "partialResult of UDAF should be 1");
+    if (partialResult[0].isNull(0)) {
+      return;
+    }
+
+    State otherState = udaf.createState();
+    Binary otherStateBinary = partialResult[0].getBinary(0);
+    otherState.deserialize(otherStateBinary.getValues());
+
+    udaf.combineState(state, otherState);
+  }
+
+  // Currently, UDAF does not support optimization
+  // by statistics in TsFile
+  @Override
+  public void addStatistics(Statistics statistics) {
+    throw new UnsupportedOperationException(getClass().getName());
+  }
+
+  // Currently, query engine won't generate
+  // Final -> Final AggregationStep for UDAF
+  @Override
+  public void setFinal(Column finalResult) {
+    throw new UnsupportedOperationException(getClass().getName());
+  }
+
+  @Override
+  public void outputIntermediate(ColumnBuilder[] columnBuilders) {
+    checkArgument(columnBuilders.length == 1, "partialResult of UDAF should be 1");
+
+    byte[] bytes = state.serialize();
```

Comment:
This method serializes the initial state into bytes[] (even if the accumulator hasn't had any values added, it will still serialize the initial state?), but in other Accumulator implementations, it checks whether an initial value has already been set, and if so, serializes the binary data; otherwise, it appends Null. Currently, it seems that this method doesn't check whether an initial value has been initialized. I wonder if this could potentially have an impact on the result.

Reply:
I think NULL value here is just to save some space. While all built-in aggregators use additional field (like `initResult`) to indicate whether the state is initialized, we can't force users to do so. For simplicity, we just serialize the whole state object into binary.

The problem is, some built-in aggregators will append NULL in columnBuilder if the state is uninitialized, and the NULL value would be skipped in `addIntermediate()` and `removeIntermediate()` method. Since UDAF won't produce NULL intermediate output, shall we remove the **skip the NULL logic** in both methods?

IMHO, if the logic is removed and UT/IT is still passed, we can consider to safely apply this change.

Reply:
If uninitialized state is needed, users shall add a field like `isInit` in their state class, and serialize the whole state object along with the field into binary, then handle this field in `combineState()` method. You can refer to this example: [UDAFSum.java](https://github.com/apache/iotdb/pull/12005/files#diff-e9b1b5efc75765bc2ddc3ad158368fdf40cc0b6a716c5c01b1a75899c849740dR111)